### PR TITLE
Server Logout Functionality + Token Invalidation + Redis integration

### PR DESCRIPTION
**Pull request by FranCogMents**
## PR Description
Provides logout functionality for the Airlock server, provides token invalidation and integrates redis as the `tokenStore` where we store invalid tokens.

## Configuration Before Testing
- Create a Base in Airtable
- Create a Redis endpoint by signing up for a hosted Redis Provider
- Obtain your Airtable api key
- Create a Table where you will store your users
    - Ensure this table has two fields `username` and `password` 
        - Both fields should be Single Line Text
- Fill out the `server/env.example` file in the root directory with the following environment variables — some have already been pre-populated (keep in mind if you want to test for plain-text passwords you will have to change `DISABLE_HASH_PASSWORD` to `true`)
    - `AIRTABLE_API_KEY`=YOUR_API_KEY
    - `AIRTABLE_BASE_ID`=YOUR_BASE_ID
    - `AIRTABLE_ENDPOINT_URL`=https://api.airtable.com/
    - `AIRTABLE_API_VERSION`=0.1.0
    - `AIRTABLE_USER_AGENT`=Airtable.js/0.7.1
    - `AIRTABLE_USER_TABLE`=NAME_OF_YOUR_USER_TABLE
    - `AIRTABLE_USERNAME_COLUMN_NAME`=NAME_OF_YOUR_USERNAME_COLUMN_NAME
    - `AIRTABLE_USER_TABLE`=NAME_OF_YOUR_USER_TABLE
    - `DISABLE_HASH_PASSWORD`=false
    - `SALT_ROUNDS`=PROVIDE_A_NUMBER
    - `CACHE_PORT=REDIS_PORT`
    - `CACHE_HOST=REDIS_ENDPOINT`
    - `CACHE_PASSWORD=REDIS_DB_PASSWORD`

- Rename `env.example` to `.env`
- Fill out Public Key + Private Key
    - In the the `/server/config` folder
    - Generate a set of public and private keys
    - Name these keys as `pub.pem` and `priv.pem`

## Running The Server
- If you have all of the configuration setup, go to the root directory of `/server` and type the following command
    - `npm run dev`
- If successful you should have a node server running on `http://localhost:4000/`

## How to Test
- Download postman (or any tool that you prefer that will allow you to simulate CRUD operations with the server)

## Testing Logout 
**note** (after posting to this route you will not be able to make queries using the same token)
- Make a `POST` request to `http://localhost:4000/:version/:base/__DANGEROUSLY__USE__TABLE__TO__LET__USERS__LOGOUT`
- Make the post request — if successful you should receive in return
     ```{
    "success": true,
    "message": "User successfully logged out"
}

## Testing Register
- Make a `POST` request to `http://localhost:4000/:version/:base/__DANGEROUSLY__USE__TABLE__TO__LET__USERS__REGISTER`
    - In the `BODY` provide a `username` and `password` in the form of JSON
        {"username": "franco",
        "password": "integrate"}
- Make the post request — if successful you should receive in return
     ```{
        "success": true,
        "token": "Jo__tAafsFuF6PU2YpbGWTolrlAjXAWNrudCaYnMrWm3RRxGz-31n721aKoWz8B7T4ofxHm_ZsxuZUoBGWm9_EeZDF_zrBnOrEij-tMi43DXOXByqskKeM5tR20Y8O8JeRygvdRYyayq5wqf5iOAxNxRINEgdB5HHLpG98B-sHQAHfs4UYPvDvCQT6XXHzEhbthYKht3Wmfg",
        "user": {
            "id": "recIOBFLPhcSwjRHQ",
            "fields": {
                "username": "franco"
            },
            "createdTime": "2019-10-14T02:40:45.000Z"
        }
    }



## Testing Login
- Make a `POST` request to `http://localhost:4000/:version/:base/__DANGEROUSLY__USE__TABLE__TO__LET__USERS__LOGIN` 
    - In the `BODY` provide the same `username` and `password` you used in testing `/register` in the form of JSON
- Make the post request — if successful you should receive in return
    ```{
        "success": true,
        "token": "eyJhbGciOiJSUzI1NiIsInR5cCI6IkpXVCJ9.eyJpZCI6InJlY2RVOVZEOU5acUIyWllDIiwiZmllbGRzIjp7InVzZXJuYW1lIjoiZnJhbmNvIn0sImNyZWF0ZWRUaW1lIjoiMjAxOS0xMC0xNFQwNDoyOTowNi4wMDBaIiwiaWF0IjoxNTcxMDI3MzYzfQ.bqB1EwGbx8-ud61aj_KDljLtYRQL1TS9o7b2e6_ATIQdNzgtFRt",
        "user": {
            "id": "recdU9VD9NZqB2ZYC",
            "fields": {
                "username": "franco"
            },
            "createdTime": "2019-10-14T04:29:06.000Z"
        }
    }

## Testing Airtable queries
- **How to Test (you will notice some of these values come from the .env file)**
    - Open up Postman
    - Create a `GET` request
        - the url should be `http://localhost:4000/v0/BASE_ID/TABLE_NAME?`
        - this is preferably the table you are using for users
    - Configure the `Headers` tab to have the following headers and values
        - `x-airtable-application-id` = BASE_ID_VALUE
        - `x-airtable-user-agent` = Airtable.js/0.7.1
        - `x-api-version` = 0.1.0
        - `authorization` = Bearer YOUR_AIRTABLE_API_KEY
        - `token` = “**This value comes from the contents of the token you received from register/login”**
    - Now you can make a `GET` request
        - If successful you should be able to get all of the contents of the table
    
```{
        "records": [
            {
                "id": "recdU9VD9NZqB2ZYC",
                "fields": {
                    "username": "franco",
                    "password": "$2b$10$HxkyufAwdukUP90GG/SepC4tOfDl1UZViHm"
                },
                "createdTime": "2019-10-14T04:29:06.000Z"
            },
            {
                "id": "recjpfmEM4FasWzP3",
                "fields": {
                    "username": "arpan",
                    "password": "$2b$10$HxkyufAwdukUP90GG/epC4tOfDl1UZV"
                },
                "createdTime": "2019-10-13T23:45:13.000Z"
            }
        ]
    }

### CODE REVIEW

File: server/controllers/authController.js [hunk 1]
```diff
@@ -114,4 +112,35 @@ module.exports = {
     const payload = { success: false };
     return res.status(422).send(payload);
   },
+  async logout(req, res) {
```
Comment (ethanlee16) [inline server/controllers/authController.js]: We should validate that the token has a valid signature before revoking it as well!

File: server/.prettierrc [hunk 2]
```diff
@@ -1,5 +1,5 @@
 {
-  "printWidth": 80,
+  "printWidth": 90,
```
[the remark this answers is not in the document]
Comment (FranCogMents) [reply server/.prettierrc:2]: the 90's were good @ethanlee16 -- but the 90 was not intentional. was not included in #11

File: server/controllers/authController.js (hunk 1 above)
Comment (FranCogMents) [reply server/controllers/authController.js:115]: good call! -- this is reflected in #11

**Comment by FranCogMents** [comment]
This PR is being closed, all comments have been addressed in the migration to Typescript #11

**Pull request closed without merging**